src/zjs_net.c

```diff
@@ -0,0 +1,1111 @@
+
+    zjs_obj_get_double(argv[0], "port", &port);
+    zjs_obj_get_string(argv[0], "host", host, 128);
+    zjs_obj_get_double(argv[0], "localPort", &localPort);
+    zjs_obj_get_string(argv[0], "localAddress", localAddress, 128);
+    zjs_obj_get_double(argv[0], "family", &fam);
+    if (fam == 0) {
+        fam = 4;
+    }
+    // TODO: get: .hints, .lookup
+
+    DBG_PRINT("port: %u, host: %s, localPort: %u, localAddress: %s, socket=%u\n", (uint32_t)port, host, (uint32_t)localPort, localAddress, this);
+
+    if (fam == 6) {
+        struct sockaddr_in6 my_addr6 = { 0 };
+
+        my_addr6.sin6_family = AF_INET6;
+        my_addr6.sin6_port = htons((uint32_t)localPort);
+
+        CHECK(net_addr_pton(AF_INET6,
+                            localAddress,
+                            &my_addr6.sin6_addr));
+        // bind to our local address
+        CHECK(net_context_bind(handle->tcp_sock,
+                               (struct sockaddr *)&my_addr6,
+                               sizeof(struct sockaddr_in6)));
+
+        struct sockaddr_in6 peer_addr6 = { 0 };
+        peer_addr6.sin6_family = AF_INET6;
+        peer_addr6.sin6_port = htons((uint32_t)port);
+
+        CHECK(net_addr_pton(AF_INET6,
+                            host,
+                            &peer_addr6.sin6_addr));
+        // set socket.connecting property == true
+        zjs_obj_add_boolean(this, true, "connecting");
+        // connect to remote
+        CHECK(net_context_connect(handle->tcp_sock,
+                                  (struct sockaddr *)&peer_addr6,
+                                  sizeof(peer_addr6),
+                                  tcp_connected,
+                                  K_NO_WAIT,
+                                  handle));
+    } else {
+        struct sockaddr_in my_addr4 = { 0 };
+
+        my_addr4.sin_family = AF_INET;
+        my_addr4.sin_port = htons((uint32_t)localPort);
+
+        CHECK(net_addr_pton(AF_INET,
+                            localAddress,
+                            &my_addr4.sin_addr));
+        // bind to our local address
+        CHECK(net_context_bind(handle->tcp_sock,
+                               (struct sockaddr *)&my_addr4,
+                               sizeof(struct sockaddr_in)));
+
+        struct sockaddr_in peer_addr4 = { 0 };
+        peer_addr4.sin_family = AF_INET;
+        peer_addr4.sin_port = htons((uint32_t)port);
+
+        CHECK(net_addr_pton(AF_INET,
+                            host,
+                            &peer_addr4.sin_addr));
+        // set socket.connecting property == true
+        zjs_obj_add_boolean(this, true, "connecting");
+        // connect to remote
+        CHECK(net_context_connect(handle->tcp_sock,
+                                  (struct sockaddr *)&peer_addr4,
+                                  sizeof(peer_addr4),
+                                  tcp_connected,
+                                  K_NO_WAIT,
+                                  handle));
```

Comment:
This API should call your callback w/ success 0 or failure < 0, so we should make the JS app retry connection instead of blocking until IF_UP and waiting a bit. (Per lunch discussion.)